src/pretix/base/datasync/datasync.py

```diff
@@ -0,0 +1,425 @@
+        self.messages = messages
+        self.full_message = full_message
+
+
+StaticMapping = namedtuple('StaticMapping', ('pk', 'pretix_model', 'external_object_type', 'pretix_pk', 'external_pk', 'property_mapping'))
+
+
+class OutboundSyncProvider:
+    max_attempts = 5
+
+    def __init__(self, event):
+        self.event = event
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    @classmethod
+    @property
+    def display_name(cls):
+        return str(cls.identifier)
+
+    """
+    Adds an order to the sync queue. May only be called on derived classes which define an "identifier" attribute.
+
+    Should be called in the appropriate signal receivers, e.g.:
+
+        @receiver(order_placed, dispatch_uid="mysync_order_placed")
+        def on_order_placed(sender, order, **kwargs):
+            MySyncProvider.enqueue_order(order, "order_placed")
+    """
+    @classmethod
+    def enqueue_order(cls, order, triggered_by, not_before=None):
+        if not hasattr(cls, 'identifier'):
+            raise TypeError('Call this method on a derived class that defines an "identifier" attribute.')
+        OrderSyncQueue.objects.create(
+            order=order,
+            sync_provider=cls.identifier,
+            triggered_by=triggered_by,
+            not_before=not_before)
+
+    @classmethod
+    def get_external_link_info(cls, event, external_link_href, external_link_display_name):
+        return {
+            "href": external_link_href,
+            "val": external_link_display_name,
+        }
+
+    @classmethod
+    def get_external_link_html(cls, event, external_link_href, external_link_display_name):
+        info = cls.get_external_link_info(event, external_link_href, external_link_display_name)
+        return make_link(info, '{val}')
+
+    """
+    Optionally override to configure a different retry backoff behavior
+    """
+    def next_retry_date(self, sq):
+        return now() + timedelta(hours=1)
+
+    """
+    Optionally override this method to exclude certain orders from sync by returning False
+    """
+    def order_valid_for_sync(self, order):
+        return True
+
+    """
+    Implementations must override this property to provide the data mappings as a list of objects.
+
+    They can return instances of the StaticMapping namedtuple defined above, or create their own
+    class (e.g. a Django model).
+
+    The returned objects must have at least the following properties:
+    - pk:                   unique identifier
+    - pretix_model:         which pretix model to use as data source in this mapping. possible values are
+                            the keys of sourcefields.AVAILABLE_MODELS
+    - external_object_type: destination object type in the target system. opaque string of maximum 128 characters.
+    - pretix_pk:            which pretix data field should be used to identify the mapped object. any
```

Comment:
Same here, I'm quite unhappy with the naming after reading the pretix-hubspot diff. This reads more like a contradiction than anything else, as it is a key, but very much does not feel like a primary key.
```
pretix_pk="attendee_or_order_email",
```

Reply:
I renamed `pretix_pk` to `pretix_id_field` and `external_pk` to `external_id_field`